### PR TITLE
Generate cancellation message for currentState=null desiredState=DROPPED

### LINKED ISSUE #830: Generate state transition cancellation message when currentState=null desiredState=DROPPED
In current version of MessageGenerationPhase, we skip message generation when currentState=null and desiredState=DROPPED to avoid infinite OFFLINE-> DROPPED messages. However, this also cause Helix fail to send cancellation message. This issue tracks the fix for this problem.

### PR DESCRIPTION
**Pull request by mgao0**
### Issues

- [x] My PR addresses the following Helix issues and references them in the PR description:

Fixes #830 

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:

To fix the issue that Helix fail to send cancel message when currentState=null and desiredState=DROPPED, this PR adds message generation logic to the condition before the "continue". To reduce the duplication of code, separate the related code section out and put into individual methods. 
Condition: current state = null, desired state = dropped, there is pending message. Current state is checked, and since it is null, it is set to the initial state OFFLINE for further handling. 
Before fix: all message generation logics are skipped for this instance, this handling is to fix a previous problem where under this condition, infinite "OFFLINE to DROPPED" state transition messages. And we are having the problem that because we skip the message generation logics, the cancellation message for the pending message won't be generated.
After fix: this fix separated the generation of cancellation message for pending message from other message generation logics to be a single method, and call this method when the above edge case is met.

### Tests

- [x] The following tests are written for this issue:

TestSendCancellationMessageFromOFFLINEToDROPPED

- [x] The following is the result of the "mvn test" command on the appropriate module:

Results :

Tests run: 1084, Failures: 0, Errors: 0, Skipped: 0

[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  01:09 h
[INFO] Finished at: 2020-03-02T11:16:06-08:00
[INFO] ------------------------------------------------------------------------

### Commits

- [x] My commits all reference appropriate Apache Helix GitHub issues in their subject lines. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation (Optional)

- [x] In case of new functionality, my PR adds documentation in the following wiki page:

(Link the GitHub wiki you added)

### Code Quality

- [x] My diff has been formatted using helix-style.xml 
(helix-style-intellij.xml if IntelliJ IDE is used)

### CODE REVIEW

File: helix-core/src/test/java/org/apache/helix/controller/stages/TestSendCancellationMessageFromOFFLINEToDROPPED.java [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+import org.testng.annotations.Test;
+
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+
+/**
+ * This test checks the cancellation message generation when currentState=null and desiredState=DROPPED
+ */
+public class TestSendCancellationMessageFromOFFLINEToDROPPED extends MessageGenerationPhase {
```
Comment (jiajunwang) [inline helix-core/src/test/java/org/apache/helix/controller/stages/TestSendCancellationMessageFromOFFLINEToDROPPED.java]: Can we just add this test method to some existing test classes? And the name is too long.

Comment (mgao0) [reply helix-core/src/test/java/org/apache/helix/controller/stages/TestSendCancellationMessageFromOFFLINEToDROPPED.java]: I'll change it to a shorter name. It has to be a separate class because it needs to extend MessageGenerationPhase so I can run this class alone. Since it is a very specific case, it's better we can run this stage alone so we can control the input to this message generation phase.

Comment (jiajunwang) [reply helix-core/src/test/java/org/apache/helix/controller/stages/TestSendCancellationMessageFromOFFLINEToDROPPED.java]: Sure, sounds good.

**Comment by mgao0** [comment]
> I haven't finish reading details yet. Overall, when we try to move code to a separate method, try to simplify the logic and parameters as much as possible. What I observed here are very long parameter lists. Can we try to simplify it?
> 
> Moreover, could you provide more details about the old behavior and the issue. Then compare it with the new one. Thanks.

Good point. Please see the updated PR description, hope it explains the issue. Please let me know how I can improve it.
Per your concerns about long parameter list, I have tried to simplify it and I think this is the best I can do right now. All the parameters are needed due to the complex logic, and the purpose of moving into a separate method is to avoid duplicated code because they will be called under two different conditions.

**Comment by mgao0** [comment]
This PR is ready to be merged, approved by @dasahcc

**Comment by dasahcc** [comment]
Please fix your author and email for commit message.
